File: algorithms/dp_patterns/go/coin_change_ways.go
package dp_patterns

type pair struct {
	sum int
	i   int
}

func CoinChangeWaysRecursive(sum int, coins []int, i int) int {
	if sum == 0 {
		return 1
	}
	if i >= len(coins) || sum < 0 {
		return 0
	}
	return CoinChangeWaysRecursive(sum-coins[i], coins, i) + CoinChangeWaysRecursive(sum, coins, i+1)
}

func CoinChangeWaysTopDownHelper(sum int, coins []int, i int, mem map[pair]int) int {
	if sum == 0 {
		return 1
	}
	if i >= len(coins) || sum < 0 {
		return 0
	}
	key := pair{sum, i}
	if _, exists := mem[key]; !exists {
		mem[key] = CoinChangeWaysRecursive(sum-coins[i], coins, i) + CoinChangeWaysRecursive(sum, coins, i+1)
	}
	return mem[key]
}

func CoinChangeWaysTopDown(sum int, coins []int, i int) int {
	mem := make(map[pair]int)
	return CoinChangeWaysTopDownHelper(sum, coins, 0, mem)
}

func CoinChangeWaysBottomUp(sum int, coins []int, i int) int {
	dp := make([]int, sum+1)
	dp[0] = 1
	for _, coin := range coins {
		for s := coin; s < sum+1; s++ {
			dp[s] = dp[s] + dp[s-coin]
		}
	}
	return dp[sum]
}

File: algorithms/dynamic_programming/python/coin_change_ways.py
class Solution:
    def count(
        self,
        coins: list[int],
        sum: int,
    ) -> int:
        dp = [0] * (sum + 1)
        dp[0] = 1
        for coin in coins:
            for i in range(coin, sum + 1):
                dp[i] += dp[i - coin]
        return dp[sum]


class SolutionRecursive:
    def count(
        self,
        coins: list[int],
        sum: int,
    ) -> int:
        def rec(i: int, sm: int) -> int:
            if sm == 0:
                return 1
            if sm < 0 or i >= len(coins):
                return 0
            return rec(i, sm - coins[i]) + rec(i + 1, sm)

        return rec(0, sum)


class SolutionTopDown:
    def count(
        self,
        coins: list[int],
        sum: int,
    ) -> int:
        def rec(i: int, sm: int) -> int:
            if sm == 0:
                return 1
            if sm < 0 or i >= len(coins):
                return 0

            key = (i, sm)
            if key not in mem:
                mem[key] = rec(i, sm - coins[i]) + rec(i + 1, sm)
            return mem[key]

        mem = {}
        return rec(0, sum)


class SolutionBottomUp:
    def count(
        self,
        coins: list[int],
        sum: int,
    ) -> int:
        dp = [0] * (sum + 1)
        dp[0] = 1
        for coin in coins:
            for i in range(coin, sum + 1):
                dp[i] += dp[i - coin]
        return dp[sum]


if __name__ == "__main__":
    coins = [2, 5, 3, 6]
    sum = 10

    # coins = [1, 2, 3]
    # sum = 4
    print(Solution().count(coins, sum))
    print(SolutionRecursive().count(coins, sum))
    print(SolutionTopDown().count(coins, sum))
    print(SolutionBottomUp().count(coins, sum))
